From d217d2914d6956680a05b6aaf7291a8ed2fc18cf Mon Sep 17 00:00:00 2001
From: "Randzio, Pawel" <pawel.randzio@intel.com>
Date: Fri, 27 May 2022 11:11:11 +0200
Subject: [PATCH] Fix tboot_dying_cpu logic to not include BSP as AP

While tracking the number of APs waiting for sipi kernel mistakenly
increments the ap_wfs_count variable (counting BSP as AP) thus
triggering a BUG_ON in suspend procedure casuing a kernel panic.

This logic fix solves that problem.

Signed-off-by: Randzio, Pawel <pawel.randzio@intel.com>
---
 arch/x86/kernel/tboot.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kernel/tboot.c b/arch/x86/kernel/tboot.c
index 0c1154a1c403..dfbd868cff9e 100644
--- a/arch/x86/kernel/tboot.c
+++ b/arch/x86/kernel/tboot.c
@@ -337,11 +337,11 @@ static int tboot_wait_for_aps(int num_aps)
 
 static int tboot_dying_cpu(unsigned int cpu)
 {
-	atomic_inc(&ap_wfs_count);
 	if (num_online_cpus() == 1) {
 		if (tboot_wait_for_aps(atomic_read(&ap_wfs_count)))
 			return -EBUSY;
-	}
+	} else
+		atomic_inc(&ap_wfs_count);
 	return 0;
 }
 
-- 
2.25.1

